### PR TITLE
feat(routing): hot reload intelligent routing overlays

### PR DESCRIPTION
**Pull request by nerdalert**
## Summary

  Adds a versioned, content-addressed routing-overlay consumer, admission-aware routing, session affinity, and in-process hot reload to intelligent_route.

  Praxis AI validates each overlay envelope, atomically installs an immutable request-time snapshot, and exposes the semantic revision it accepted and served. In-flight requests continue using the snapshot they started with while later requests observe the new revision.

  Grid is the first producer of this contract, but the Praxis configuration, filter names, headers, metadata, and wire format are intentionally producer-neutral. The consumer does not depend on Kubernetes APIs, Grid APIs, SWIM, CRDTs, or a specific operator at request time.

  ## Operating Modes

  intelligent_route supports two mutually exclusive configuration modes:

  - Static mode: Candidates and local_site are declared inline in the Praxis YAML configuration and remain immutable for the life of the filter.
  - Overlay mode: Candidates and local_site are loaded from a routing-overlay file and can be replaced without restarting Praxis.

  Configuration fails when both modes are configured or when neither mode provides routing state.

  Example overlay mode:

```yaml
  - filter: intelligent_route
    overlay_file: /etc/praxis/routing/routing-overlay.json
    model_header: X-Model
    expected_overlay_scope:
      network: production
      gateway: east-edge
      namespace: praxis-system
      local_site: east
    provider_hop_clusters:
      - east-provider
      - west-provider
    reload:
      enabled: true
      debounce_ms: 500
    session_affinity:
      enabled: true
      header: X-Session-Id
      ttl_secs: 3600
```

  ## Revision Lifecycle

```
  producer renders revision A
    -> distribution layer publishes revision A
    -> Praxis validates and accepts revision A
    -> intelligent_route serves requests from revision A
```

  The stages have distinct meanings:

  - Rendered: The producer constructed a valid envelope and calculated its semantic revision.
  - Distributed: The envelope reached the configuration volume mounted into the Praxis deployment.
  - Accepted: Praxis parsed the envelope, verified its digest, checked its scope, validated every candidate, and installed the snapshot.
  - Serving: A request selected a route from that exact immutable snapshot.

  Keeping these stages separate makes it possible to distinguish producer failures, distribution delays, consumer rejection, and stale serving state.

  ## Versioned Envelope Contract

  The supported v1 envelope uses schema version 1.0.0 and contains:

  - A content-addressed semantic revision.
  - A SHA-256 content digest.
  - Producer-neutral scope.
  - Bounded provenance.
  - The ordered routing payload.

  The envelope shape is:

```json
  {
    "schema_version": "1.0.0",
    "revision": {
      "kind": "content_addressed",
      "algorithm": "sha256",
      "value": "<64 lowercase hexadecimal characters>"
    },
    "content_digest": {
      "algorithm": "sha256",
      "value": "<same digest as revision.value>"
    },
    "scope": {
      "network": "production",
      "gateway": "east-edge",
      "namespace": "praxis-system",
      "local_site": "east"
    },
    "provenance": {
      "producer": "routing-controller",
      "producer_version": "v1.0.0",
      "source_name": "production",
      "source_uid": "opaque-source-identifier",
      "source_generation": 42,
      "rendered_at": "2026-07-31T00:00:00Z"
    },
    "overlay": {
      "network": "production",
      "local_site": "east",
      "candidates": []
    }
  }
```

  ## Envelope Validation

  Praxis validates the complete envelope before installing a snapshot:

  - schema_version must be exactly 1.0.0.
  - Revision kind must be content_addressed.
  - Revision and digest algorithms must be sha256.
  - Revision and digest values must be 64 lowercase hexadecimal characters.
  - revision.value and content_digest.value must agree.
  - Praxis recomputes the RFC 8785 canonical SHA-256 digest over the routing payload.
  - Envelope scope must be internally consistent with the embedded overlay.
  - Provenance source information must be structurally valid and consistent with scope.
  - Source generation must be valid.
  - Scope and provenance strings are bounded.
  - Candidate identity, admission state, locality, rank, stable ID, and credential references are validated.
  - Unknown admission states fail closed instead of defaulting to an eligible state.
  - Candidate stable_id and selection_tier values must be non-empty and bounded when present.

  An envelope is accepted only after all validation succeeds. Partial state is never installed.

  ## Expected Scope

  Deployments can configure any combination of expected network, gateway, namespace, and local-site values.

  Every configured field is checked during initial load and every reload. A mismatch rejects the new envelope and preserves the existing serving snapshot.

  expected_overlay_scope is valid only in envelope-backed overlay mode. When expected scope is configured, legacy payloads are rejected so a caller cannot bypass scope validation by removing envelope metadata.

  ## Strict Format Detection

  Format detection is deterministic and fail closed.

  The presence of schema_version or any other reserved envelope field selects the envelope path. Once the document is identified as an envelope, malformed or incomplete content cannot fall back to legacy parsing.

  This prevents an invalid, partially written, or intentionally downgraded envelope from being accepted as an unscoped legacy payload.

  Legacy payloads remain available only when strict expected scope is not configured.

  ## Candidate Contract

  Overlay candidates can carry:

  - Capability kind and name.
  - Owning site and upstream cluster.
  - Stable candidate identity.
  - Admission state.
  - Producer-assigned rank.
  - Producer-assigned selection tier.
  - Freshness.
  - Credential reference.
  - Additional forward-compatible metadata.

  Candidate identity includes (kind, name, site, cluster), allowing multiple independent provider clusters to offer the same model at one site.

  The configuration producer owns health, capacity, geography, cost, and other scoring decisions. It renders candidates in preferred order. Praxis preserves that order and applies request local eligibility and session rules rather than recreating the producer’s scoring engine.

  ## Admission-Aware Routing

  The overlay supports three admission states:

  - new_and_existing: Eligible for new requests and existing bound sessions.
  - existing_only: Ineligible for new sessions but remains eligible for sessions already bound to the candidate.
  - none: Excluded from routing entirely.

  When session affinity is disabled or no valid session key is present, every request is treated as new and only new_and_existing candidates are eligible.

  This allows a producer to drain a provider without immediately breaking eligible sessions already using it.

  ## Session Affinity

  Optional session affinity binds a request key to a candidate stable_id.

  The key can be extracted from a configured header or cookie. Subsequent requests reuse the bound candidate while:

  - The binding has not expired.
  - The candidate remains present in the current snapshot.
  - The candidate is not excluded.
  - The requested capability still matches.

  If the binding expires or the candidate disappears or becomes excluded, Praxis selects another eligible candidate and replaces the binding.

  Session-affinity state is deliberately bounded:

  - Disabled by default.
  - Stored in process memory.
  - Limited to 10,000 bindings.
  - Default TTL is 3,600 seconds.
  - Maximum TTL is 86,400 seconds.
  - Expired entries are evicted before capacity rejection.
  - Bindings are local to one Praxis process and are lost on restart.

  Shared or restart-persistent affinity remains outside this PR.

  ## Hot Reload

  When overlay reload is enabled, Praxis:

  - Reads and validates the initial overlay before filter construction succeeds.
  - Watches the overlay file’s parent directory rather than only the file inode.
  - Detects Kubernetes projected-volume ..data symlink replacement.
  - Debounces related filesystem events.
  - Re-reads the complete file after an event.
  - Rejects files larger than 2 MiB.
  - Hashes the raw content and skips unchanged files.
  - Parses and validates the complete replacement.
  - Atomically swaps the accepted snapshot through ArcSwap.
  - Lets in-flight requests continue using their previously loaded snapshot.
  - Cancels and joins the watcher through a bounded shutdown path.

  Watcher initialization uses a readiness rendezvous. Filter construction does not report success until the background watcher has registered successfully.

  The watcher also performs an authoritative re-read after registration to close the race between the initial file read and watcher startup.

  ## Kubernetes Projected Volumes

  The watcher supports the normal Kubernetes ConfigMap projected-volume update mechanism.

  The ConfigMap must be mounted as a directory. It must not use a subPath mount because subPath bypasses Kubernetes’ atomic ..data symlink replacement and prevents the running process from
  observing updates.

  No Kubernetes API permission is required by Praxis. Kubernetes only projects the file; Praxis watches and validates the local filesystem.

  ## Cold Start And Last-Known-Good Behavior

  Cold start fails closed:

  - Missing initial file fails filter construction.
  - Unreadable initial file fails filter construction.
  - Oversized initial file fails filter construction.
  - Invalid JSON fails filter construction.
  - Unsupported schema fails filter construction.
  - Digest, scope, provenance, or candidate validation failure fails filter construction.
  - A legacy payload fails construction when strict envelope scope is required.

  Reloads preserve availability:

  - Invalid JSON retains the current snapshot.
  - Missing or unreadable files retain the current snapshot.
  - Oversized files retain the current snapshot.
  - Unsupported schema retains the current snapshot.
  - Digest mismatch retains the current snapshot.
  - Scope or provenance mismatch retains the current snapshot.
  - Invalid candidate state retains the current snapshot.
  - Attempted legacy downgrade retains the current snapshot when strict envelope scope is enabled.

  Last-known-good state is same-process memory. This PR does not persist it across process or pod restarts.

  ## Request-Path Behavior

  The request path loads one immutable snapshot and uses that snapshot for the complete routing decision.

  It does not:

  - Read the overlay file.
  - Query Kubernetes.
  - Query the configuration producer.
  - Query Grid, SWIM, or CRDT state.
  - Recalculate health or capacity scoring.
  - Hold a lock across candidate selection.
  - Observe a partially installed update.

  Candidate selection, serving revision, route metadata, session behavior, and provider-hop context are derived from the same snapshot.

  ## Provider-Hop Context

  For clusters explicitly listed in provider_hop_clusters, intelligent_route emits bounded routing context for an authenticated Praxis provider gateway:

  - x-ai-routing-candidate: The selected candidate’s stable ID.
  - x-ai-routing-request-id: A generated correlation identifier for the provider hop.
  - x-ai-routing-revision: The serving envelope revision.

  Caller-supplied values in this namespace are stripped before trusted replacements are created.

  The context is emitted only for allow-listed provider-hop clusters. Direct provider APIs and ordinary backend clusters should not appear in provider_hop_clusters.

  These headers are routing and correlation context, not authentication credentials. #386 requires downstream mTLS and peer_identity_trust before the provider gateway consumes them.

  Credential values are never included. Credential references remain typed overlay data for the later credential_inject boundary.

  Legacy and static snapshots do not emit an overlay revision because they do not have a verified semantic revision.

  ## Metadata And Observability

  Successful routing records bounded metadata under the intelligent_route.* namespace, including:

  - Capability kind and name.
  - Selected site and cluster.
  - Local site.
  - Stable candidate ID.
  - Admission state.
  - Optional rank and selection tier.
  - Session-bound, reused, and failover state when affinity is enabled.

  Reload events report accepted revision, serving revision, and previous serving revision. Rejected reloads report the retained serving revision without logging request payloads, credentials,
  or secret values.

  ## Forward Compatibility And Secret Safety

  Unknown additive fields are accepted in envelope, provenance, and candidate objects so producers can add non-breaking metadata before every consumer is upgraded.

  Credential objects intentionally reject unknown fields. This prevents token-like fields or inline secret material from silently entering a routing overlay.

  The overlay may contain credential references, but it must never contain credential values.

  ## Runtime Configuration Boundary

  Hot reload replaces only:

  - Candidate state.
  - Candidate order and admission state.
  - Local-site identity.
  - Semantic revision and related envelope metadata.

  Hot reload does not:

  - Create or delete load_balancer clusters.
  - Change cluster endpoints.
  - Change TLS configuration.
  - Change peer trust policy.
  - Add credential values.
  - Change listener configuration.
  - Rebuild the full filter pipeline.

  Every cluster that may appear in an overlay must already exist in the downstream load_balancer configuration. An unknown cluster can be accepted as candidate data but will fail when the
  request reaches cluster resolution; validating candidate clusters against the load-balancer catalog remains a separate pipeline-level concern.

  ## Scope

  This PR does not add:

  - Provider-local route authorization.
  - Provider credential resolution or injection.
  - Dynamic upstream cluster creation.
  - Persistent last-known-good storage.
  - Shared session storage.
  - Cross-instance session affinity.
  - A second provider scoring engine.
  - Kubernetes watches or API clients inside Praxis.
  - General-purpose control-plane functionality in Praxis core.

  Provider-local authorization, trust enforcement, and final-hop credential replacement are added by #386.

  ## Included Coverage

  The PR includes coverage for:

  - Accepted v1 envelopes.
  - Canonical digest verification.
  - Revision and digest disagreement.
  - Unsupported schema versions.
  - Scope and provenance mismatch.
  - Malformed and partial envelope downgrade attempts.
  - Forward-compatible additive fields.
  - Legacy compatibility and strict-scope rejection.
  - Cold-start failures.
  - Invalid-reload last-known-good retention.
  - Unchanged-content reload suppression.
  - Kubernetes-style symlink swaps.
  - Watcher startup failure and startup race handling.
  - Watcher cancellation and bounded cleanup.
  - Candidate admission behavior.
  - Same-site multiple-provider candidates.
  - Session creation, reuse, drain, expiry, failover, and capacity bounds.
  - Provider-hop header stripping, generation, allow-listing, and revision propagation.
  - Exact accepted fixture revisions.
  - Cross-repository contract fixtures.

  ## Demonstration

  The complete rendered, distributed, accepted, and serving revision chain is exercised by the Grid global ingress demo (https://github.com/praxis-proxy/grid/tree/main/demos/grid-glb-demo).

  The demo also proves:

  - Hot reload without replacing or restarting the Praxis edge pod.
  - Multiple provider candidates for the same model and site.
  - Metrics-driven provider drain.
  - Existing-session preservation during drain.
  - Invalid-overlay last-known-good retention.
  - Cold invalid-overlay failure.
  - Serving-revision propagation through the provider hop.

  ## Validation

  - make lint
  - cargo check --workspace
  - cargo test -p praxis-ai-filters register
  - Workspace tests completed during stack validation.
  - The final stacked image passed the cold quick Grid demo with all applicable runtime proof points.

  ## Review Focus

  - Canonical digest computation and content-addressed revision semantics.
  - Strict envelope detection and downgrade resistance.
  - Expected-scope and provenance validation.
  - Projected-volume watcher startup and update races.
  - Atomic snapshot replacement and in-flight request consistency.
  - Cold-start versus reload failure behavior.
  - Admission-state and session-affinity semantics.
  - Provider-hop allow-listing and spoofed-header removal.
  - Credential-reference handling without secret material.
  - The boundary between reloadable candidate state and static proxy configuration.

### CODE REVIEW

File: examples/README.md [hunk 1]
```diff
@@ -15,6 +15,13 @@ before sending requests.
...
+### Grid Routing
+
+| File | Description |
+| ------ | ------------- |
+| [grid-route-inference.yaml](configs/grid-route-inference.yaml) | Routes requests to different upstream clusters based on the inference model name extracted from a configured request header |
+| [grid-route-mcp.yaml](configs/grid-route-mcp.yaml) | Routes MCP `tools/call` requests to the cluster that owns the requested tool using `mcp.name` metadata |
+
 ### General
 
 | File | Description |
```
[the remark this answers is not in the document]
Comment (nerdalert) [reply examples/README.md:20]: Addressed. The renamed `intelligent-route-overlay.yaml` example is listed in the examples table.

File: filters/src/grid/descriptor.rs [hunk 2]
```diff
@@ -0,0 +1,521 @@
...
+
+// -----------------------------------------------------------------------------
+// AdmissionState
+// -----------------------------------------------------------------------------
+
+/// Grid-operator-assigned admission state for a routing candidate.
+///
+/// Controls whether a candidate accepts new sessions, existing
+/// sessions only, or is excluded from routing entirely.  Unknown
+/// values from the overlay default to [`NewAndExisting`] for
```
[the remark this answers is not in the document]
Comment (nerdalert) [reply filters/src/grid/descriptor.rs]: Addressed. The docs now describe fail-closed rejection and last-known-good retention.

File: filters/src/routing/descriptor.rs [hunk 3]
```diff
@@ -0,0 +1,521 @@
...
+    fn same_name_different_site_not_duplicate() {
+        let result = validate_candidates(vec![
+            candidate("inference_model", "llama", "site-a", "c1"),
+            candidate("inference_model", "llama", "site-b", "c2"),
+        ]);
+        assert!(result.is_ok());
+    }
+
+    #[test]
+    fn same_model_different_site_not_duplicate() {
```
[the remark this answers is not in the document]
Comment (nerdalert) [reply filters/src/routing/descriptor.rs]: Fixed. Removed the duplicate test while retaining equivalent coverage.

File: filters/src/routing/descriptor.rs [hunk 4]
```diff
@@ -120,14 +197,31 @@ pub(crate) struct RouteCandidate {
...
 
 // -----------------------------------------------------------------------------
 // Validation
 // -----------------------------------------------------------------------------
 
+/// Build a deterministic stable ID from candidate identity fields.
+///
+/// Used as a fallback when the overlay does not provide an
+/// explicit `stable_id`.
+pub(crate) fn default_stable_id(kind: CapabilityKind, name: &str, site: &str, cluster: &str) -> Arc<str> {
```
Comment (aslakknutsen) [inline filters/src/routing/descriptor.rs:222]: "When overlay candidates omit stable_id, default_stable_id concatenates kind/name/site/cluster (descriptor.rs:221-223). With bounded 256-byte identity fields the composite can reach ~786 bytes, exceeding MAX_HEADER_VALUE_LEN (256). Overlays accept at load time, but write_provider_context fails at request time for allowlisted provider_hop_clusters after cluster/metadata mutation (intelligent_route.rs:671-680, 973-977). Producers in envelope mode typically supply explicit stable_id (fixtures always do), so operational risk is low; still worth bounding or validating the fallback at accept time."

Comment (nerdalert) [reply filters/src/routing/descriptor.rs:222]: Ty! Updated the fallback to preserve the readable composite when it is a valid value within 256 bytes, and otherwise generate a deterministic sha256:<digest> ID. The hash uses length-delimited identity fields to preserve field boundaries. I also added regression tests for maximum-length fields and ambiguous composite inputs.

**Comment by shaneutt** [comment]
</img>

**Comment by shaneutt** [comment]
</img>

**Comment by shaneutt** [comment]
</img>

**Review by shaneutt — APPROVED** [review]
